File: mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerRunReport.java
package com.mifos.api.datamanager;

import com.mifos.api.BaseApiManager;
import com.mifos.objects.group.CenterInfo;
import com.mifos.objects.runreports.FullParameterListResponse;
import com.mifos.objects.runreports.client.ClientReportTypeItem;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

/**
 * Created by Rajan Maurya on 05/02/17.
 */
@Singleton
public class DataManagerRunReport {

    public final BaseApiManager mBaseApiManager;

    @Inject
    public DataManagerRunReport(BaseApiManager baseApiManager) {
        mBaseApiManager = baseApiManager;
    }

    public Observable<List<ClientReportTypeItem>> getReportCategories(String reportCategory,
                                                                      boolean genericResultSet,
                                                                      boolean parameterType) {
        return mBaseApiManager.getRunReportsService().getReportCategories(reportCategory,
                genericResultSet, parameterType);
    }

    public Observable<FullParameterListResponse> getReportFullParameterList(
            String reportName, boolean parameterType) {
        return mBaseApiManager.getRunReportsService()
                .getReportFullParameterList(reportName, parameterType);
    }

    public Observable<FullParameterListResponse> getReportParameterDetails(
            String parameterName, boolean parameterType) {
        return mBaseApiManager.getRunReportsService()
                .getReportParameterDetails(parameterName, parameterType);
    }

    public Observable<FullParameterListResponse> getRunReportWithQuery(
            String reportName, Map<String, String> options) {
        return mBaseApiManager.getRunReportsService()
                .getRunReportWithQuery(reportName, options);
    }

    public Observable<List<CenterInfo>> getCenterSummarInfo(int centerId,
                                                            boolean genericResultSet) {
        return mBaseApiManager.getRunReportsService()
                .getCenterSummaryInfo(centerId, genericResultSet);
    }

    public Observable<FullParameterListResponse> getRunReportOffices(
            String parameterName, int officeId, boolean parameterType) {
        return mBaseApiManager.getRunReportsService().
                getReportOffice(parameterName, officeId, parameterType);
    }

    public Observable<FullParameterListResponse> getRunReportProduct(
            String parameterName, String currency, boolean parameterType) {
        return mBaseApiManager.getRunReportsService().
                getReportProduct(parameterName, currency, parameterType);
    }
}

File: core/network/src/main/java/com/mifos/core/network/datamanager/DataManagerRunReport.kt
package com.mifos.core.network.datamanager

import com.mifos.core.network.BaseApiManager
import com.mifos.core.objects.group.CenterInfo
import com.mifos.core.objects.runreports.FullParameterListResponse
import com.mifos.core.objects.runreports.client.ClientReportTypeItem
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Rajan Maurya on 05/02/17.
 */
@Singleton
class DataManagerRunReport @Inject constructor(val mBaseApiManager: BaseApiManager) {
    suspend fun getReportCategories(
        reportCategory: String?,
        genericResultSet: Boolean,
        parameterType: Boolean
    ): List<ClientReportTypeItem> {
        return mBaseApiManager.runReportsService.getReportCategories(
            reportCategory,
            genericResultSet, parameterType
        )
    }

    suspend fun getReportFullParameterList(
        reportName: String, parameterType: Boolean
    ): FullParameterListResponse {
        return mBaseApiManager.runReportsService
            .getReportFullParameterList(reportName, parameterType)
    }

    suspend fun getReportParameterDetails(
        parameterName: String, parameterType: Boolean
    ): FullParameterListResponse {
        return mBaseApiManager.runReportsService
            .getReportParameterDetails(parameterName, parameterType)
    }

    suspend fun getRunReportWithQuery(
        reportName: String, options: Map<String, String>
    ): FullParameterListResponse {
        return mBaseApiManager.runReportsService
            .getRunReportWithQuery(reportName, options)
    }

    suspend fun getCenterSummaryInfo(
        centerId: Int,
        genericResultSet: Boolean
    ): List<CenterInfo> {
        return mBaseApiManager.runReportsService
            .getCenterSummaryInfo(centerId, genericResultSet)
    }

    suspend fun getRunReportOffices(
        parameterName: String, officeId: Int, parameterType: Boolean
    ): FullParameterListResponse {
        return mBaseApiManager.runReportsService.getReportOffice(
            parameterName,
            officeId,
            parameterType
        )
    }

    suspend fun getRunReportProduct(
        parameterName: String, currency: String, parameterType: Boolean
    ): FullParameterListResponse {
        return mBaseApiManager.runReportsService.getReportProduct(
            parameterName,
            currency,
            parameterType
        )
    }
}